src/hotspot/share/gc/shenandoah/shenandoahGeneration.cpp

```diff
@@ -882,43 +882,47 @@ size_t ShenandoahGeneration::increment_affiliated_region_count() {
   // During full gc, multiple GC worker threads may change region affiliations without a lock.  No lock is enforced
   // on read and write of _affiliated_region_count.  At the end of full gc, a single thread overwrites the count with
   // a coherent value.
-  _affiliated_region_count++;
-  return _affiliated_region_count;
+  return Atomic::add(&_affiliated_region_count, (size_t) 1);
 }
 
 size_t ShenandoahGeneration::decrement_affiliated_region_count() {
-  shenandoah_assert_heaplocked_or_safepoint();
+  // Assertions only hold true for Java threads since they call this method under heap lock.
+  bool const is_java_thread = Thread::current()->is_Java_thread();
```

Comment:
It will be weird if I only change   `decrement_affiliated_region_count` to `decrement_affiliated_region_count_under_lock` in this file, while all the other `decrement_x` / `decrease_x` /`increment_x` / `increase_x` methods in files follow the same convention.

It is probably better to add new one like `decrement_affiliated_region_count_without_lock` and not change the existing methods's behavior.